src/ucp/core/ucp_mm.c

```diff
@@ -105,23 +106,20 @@ ucs_status_t ucp_mem_rereg_mds(ucp_context_h context, ucp_md_map_t reg_md_map,
         } else if (!length) {
             /* don't register zero-length regions */
             continue;
-        } else if ((md_attr->cap.flags & UCT_MD_FLAG_REG) &&
-                   (md_attr->cap.reg_mem_types & UCS_BIT(mem_type))) {
+        } else if (md_attr->cap.flags & UCT_MD_FLAG_REG) {
+            if (!(md_attr->cap.reg_mem_types & UCS_BIT(mem_type))) {
+                status = new_md_map ? UCS_OK : UCS_ERR_UNSUPPORTED;
```

Comment:
why need to check new_md_map ?
also, if UCS_ERR_UNSUPPORTED, shouldn't we exit the loop?
seems we can register some MDs and still return failure?

Reply:
we should completely fail if mem reg is not supported by at least one md, and undo any registrations. otherwise it's a confusing definition of this function's behavior. Just as we were doing before this PR, when one of the MD's failed to register.
As we discussed in f2f, we can let protocols such as rndv fallback to bcopy in case of memreg failure, for **any** reason.

Reply:
we have some places where we just try to register on all available MDs. 
`ucp_mem_alloc` for instance. What would be the correct behavior there? Check for allocated memory type and construct MD map accordingly?

Reply:
yes. AFAIK today protocols assume that all registrations work if rereg_mds returns UCS_OK - because they try to use the lanes for zcopy. We can't just return UCS_OK if smth is not registered.
Maybe add param "int force" to rereg_mds to define if partial registration should be OK.

Reply:
Looks like we will have to pass this force flag always then. 
 I'll double check the flows